src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs

```diff
@@ -190,49 +205,165 @@ private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
                 }
-
-                // Ignore parenthesized operations
-                IOperation? ancestor = caseChangingInvocation.WalkUpParentheses().WalkUpConversion().Parent;
-
-                IInvocationOperation diagnosableInvocation;
-                if (ancestor is IInvocationOperation invocationAncestor)
-                {
-                    diagnosableInvocation = invocationAncestor;
-                }
-                else if (ancestor is IArgumentOperation argumentAncestor && argumentAncestor.Parent is IInvocationOperation argumentInvocationAncestor)
+                else if (context.Operation is IBinaryOperation binaryOperation)
                 {
-                    diagnosableInvocation = argumentInvocationAncestor;
-                }
-                else
-                {
-                    return;
+                    AnalyzeBinaryOperation(context, binaryOperation, stringType);
                 }
+            },
+                OperationKind.Invocation, // a.ToLower().Method()
+                OperationKind.Binary // a.ToLower() == b.ToLower()
+            );
+        }
 
-                IMethodSymbol diagnosableMethod = diagnosableInvocation.TargetMethod;
+        private static void AnalyzeInvocation(OperationAnalysisContext context, IInvocationOperation caseChangingInvocation, INamedTypeSymbol stringType,
+            IMethodSymbol containsStringMethod, IMethodSymbol startsWithStringMethod, IMethodSymbol compareToStringMethod,
+            IMethodSymbol indexOfStringMethod, IMethodSymbol indexOfStringInt32Method, IMethodSymbol indexOfStringInt32Int32Method)
+        {
+            if (!IsOffendingMethod(caseChangingInvocation, stringType, out string? offendingMethodName))
+            {
+                return;
+            }
 
-                if (diagnosableMethod.Equals(containsStringMethod) ||
-                    diagnosableMethod.Equals(startsWithStringMethod) ||
-                    diagnosableMethod.Equals(indexOfStringMethod) ||
-                    diagnosableMethod.Equals(indexOfStringInt32Method) ||
-                    diagnosableMethod.Equals(indexOfStringInt32Int32Method))
-                {
-                    context.ReportDiagnostic(diagnosableInvocation.CreateDiagnostic(RecommendCaseInsensitiveStringComparisonRule, diagnosableMethod.Name));
-                }
-                else if (diagnosableMethod.Equals(compareToStringMethod))
+            if (!TryGetInvocationWithoutParentheses(caseChangingInvocation, out IInvocationOperation? diagnosableInvocation))
+            {
+                return;
+            }
+
+            string caseChangingApproach = GetCaseChangingApproach(offendingMethodName);
+
+            ImmutableDictionary<string, string?> dict = new Dictionary<string, string?>()
+            {
+                { CaseChangingApproachName, caseChangingApproach }
+            }.ToImmutableDictionary();
```

Comment:
NIT:
```suggestion
           ImmutableDictionary<string, string> dict = ImmutableDictionary.CreateRange(
                new Dictionary<string, string> { { CaseChangingApproachName, caseChangingApproach } });
```

Reply:
Is that actually better? I am still allocating a dictionary with the suggestion, so I'm not sure if there's an improvement.
BTW the second typeparam should be a `string?`, otherwise it complains.

Reply:
> I'm not sure if there's an improvement.

I would expect it to be better, though do not know for sure, that is why I put NIT